cmd/speedtest/speedtest.go

```diff
@@ -0,0 +1,193 @@
+var speedtestCmd = &ffcli.Command{
+	Name:       "speedtest",
+	ShortUsage: "speedtest {server|client} ...",
+	ShortHelp:  "Run a speed test",
+	Subcommands: []*ffcli.Command{
+		speedtestServerCmd,
+		speedtestClientCmd,
+	},
+	Exec: func(context.Context, []string) error {
+		return errors.New("subcommand required; run 'tailscale speedtest -h' for details")
+	},
+}
+
+// speedtestServerCmd takes necessary info like the port to
+// listen on and then passes them to the StartServer function in the speedtest package.
+// if the localhost flag is given, the server will use 127.0.0.1, otherwise the server will
+// use the tailscale ip address
+var speedtestServerCmd = &ffcli.Command{
+	Name:       "server",
+	ShortUsage: "speedtest server --host <host> --port <port> [--max-conns <max connections>]",
+	ShortHelp:  "Start a speed test server",
+	Exec:       runServer,
+	FlagSet: (func() *flag.FlagSet {
+		fs := flag.NewFlagSet("server", flag.ExitOnError)
+		fs.IntVar(&serverArgs.port, "port", 0, "port to listen on")
+		fs.BoolVar(&serverArgs.localhost, "localhost", false, "use localhost or tailscale ip")
+		fs.IntVar(&serverArgs.maxConnections, "max-conns", 1, "max number of concurrent connections allowed")
+		return fs
+	})(),
+}
+
+// speedtestClientCmd takes info like the type of test to run, message size, test time, and the host and port
+// of the speedtest server and passes them to the StartClient function in the speedtest package.
+var speedtestClientCmd = &ffcli.Command{
+	Name:       "client",
+	ShortUsage: "speedtest client {--d|--u} --host <host> --port <port> [--inc <increment>] [--seconds <number of seconds>]",
+	ShortHelp:  "Start a speed test client and connect to a speed test server",
+	Exec:       runClient,
+	FlagSet: (func() *flag.FlagSet {
+		fs := flag.NewFlagSet("client", flag.ExitOnError)
+		fs.StringVar(&clientArgs.host, "host", "", "The ip address for the speedtest server being used")
+		fs.StringVar(&clientArgs.port, "port", "", "The port of the speedtest server being used")
+		fs.IntVar(&clientArgs.inc, "inc", 1, "The increment for displaying speedtest info")
+		fs.BoolVar(&clientArgs.download, "d", false, "Include this to run a download test")
+		fs.BoolVar(&clientArgs.upload, "u", false, "Include this to run an upload test")
+		fs.IntVar(&clientArgs.seconds, "seconds", speedtest.DefaultNumSeconds, "The duration of the speed test in seconds")
+		return fs
+	})(),
+}
+
+var serverArgs struct {
+	port           int
+	localhost      bool
+	maxConnections int
+}
+
+// runServer takes the port from the serverArgs variable, finds the tailscale ip if needed, then passes them
+// to speedtest.GetListener. The listener is then passed to speedtest.StartServer. No channel is passed to
+// StartServer, because to kill the server all the user has to do is do Ctrl+c.
+func runServer(ctx context.Context, args []string) error {
+	if serverArgs.port == 0 {
+		return errors.New("port needs to be provided")
```

Comment:
If no port is provided you can listen on port 0, which will choose an available port. You can then find out what port it picked:
```
         l, err := net.Listen("tcp4", "localhost:0")
         defer l.Close()
         port := l.Addr().(*net.TCPAddr).Port
```
You'd need to print this on the console, so the person can type that port number as an argument to the client.